src/dmd/dsymbolsem.d

```diff
@@ -5989,3 +5986,95 @@ void aliasSemantic(AliasDeclaration ds, Scope* sc)
             ScopeDsymbol.multiplyDefined(Loc.initial, sx, ds);
     }
 }
+
+/**
+ * build a shim function around the compound dtor that accepts an argument
+ *  that is used to implement the deleting C++ destructor
+ *
+ * Params:
+ *  cldev = the class that contains the destructor to wrap
+ *  sc = the scope in which to analyze the new function
+ *
+ * Returns:
+ *  the shim destructor, semantically analyzed and added to the class as a member
+ */
+private FuncDeclaration buildWindowsCppDtor(ClassDeclaration cldec, Scope* sc)
+{
+    auto dtor = cldec.dtor;
+    if (!dtor)
+        return null;
+
+    Parameter param = new Parameter(STC.undefined_, Type.tuns32, Identifier.idPool("del"), null);
+    Parameters* params = new Parameters;
+    (*params).push(param);
+    auto ftype = new TypeFunction(params, Type.tvoidptr, false, LINK.cpp, dtor.storage_class);
+    auto func = new DtorDeclaration(dtor.loc, dtor.loc, dtor.storage_class, Id.cppdtor);
+    func.type = ftype;
+    if (dtor.fbody)
+    {
+        Loc loc = dtor.loc;
+        auto stmts = new Statements;
+        auto call = new CallExp(loc, dtor, null);
+        call.directcall = true;                                                                   // Class.__dtor();
+        stmts.push(new ExpStatement(loc, call));                                                  // TODO: free mem if del != 0
+        stmts.push(new ReturnStatement(loc, new CastExp(loc, new ThisExp(loc), Type.tvoidptr)));  // return cast(void*)this;
+        func.fbody = new CompoundStatement(loc, stmts);
+    }
+    func.generated = true;
+
+    sc = sc.push();
+    sc.stc &= ~STC.static_; // not a static destructor
+    sc.linkage = LINK.cpp;
+
+    cldec.members.push(func);
+    func.addMember(sc, cldec);
+    func.dsymbolSemantic(sc);
+
+    sc.pop();
+    return func;
+}
+
+/**
+ * build a shim function around the compound dtor that translates
+ *  a C++ destructor to a destructor with extern(D) calling convention
+ *
+ * Params:
+ *  cldec = the class that contains the destructor to wrap
+ *  sc = the scope in which to analyze the new function
+ *
+ * Returns:
+ *  the shim destructor, semantically analyzed and added to the class as a member
+ */
+private DtorDeclaration buildExternDDtor(ClassDeclaration cldec, Scope* sc)
+{
+    auto dtor = cldec.dtor;
+    if (!dtor)
+        return null;
+
+    if (cldec.classKind != ClassKind.cpp || !global.params.isWindows || global.params.is64bit)
+        return dtor;
+
+    // the only ABI known to be incompatible is Windows/x86
```

Comment:
Should this apply to 64bit too?
`cldec.dtor` receives an argument either way. That it's not passed on the stack won't lead to a crash, but if `dtor` takes the argument and decides to call `operator delete` and the calling function didn't supply the arg (just whatever was in RDX), then won't it just randomly delete?

Reply:
Sorry, I think you've rearranged it further than I realised...

Reply:
Yeah, `cldec.dtor` no longer receives the argument, only he function in the vtable does. If D destroys a class/struct, it always calls the version without argument.

Reply:
So `xdtor` is the one without the argument? And the one with the argument is the one that's mangled and added to the vtable, but not callable from D?

Reply:
Yes.